SWDEV-546303 - implement hipKernelGetAttribute API

sluzynsk-amd opened the pull request:
## Motivation
Resolves SWDEV-546303

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details
adds api implementation

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan
local testing and new tests

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result
new unit tests pass

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

sluzynsk-amd committed:
SWDEV-546303 - implement hipKernelGetAttribute API

Signed-off-by: Sebastian Luzynski <Sebastian.Luzynski@amd.com>

g-h-c commented on projects/clr/hipamd/src/hip_library.cpp:
```diff
@@ -177,4 +178,62 @@ hipError_t hipLibraryGetKernel(hipKernel_t* kernel, hipLibrary_t library, const
+
+  const auto* const device = hip::getCurrentDevice()->devices()[dev];
+  const auto* const wrkGrpInfo = d_kernel->getDeviceKernel(*device)->workGroupInfo();
+  if (wrkGrpInfo == nullptr) {
```
I think this check is unnecessary, the value returned by `wrkGrpInfo()` is always distinct from `nullptr`; it is a member of the object returned by `d_kernel->getDeviceKernel(*device)`

Checking `d_kernel->getDeviceKernel(*device) == nullptr` makes more sense to me (if anything)

sluzynsk-amd replied:
I agree with you, however, looking just at the function declaration it returns pointer. In my opinion `wrkGrpInfo()` should return a reference instead of a pointer. I removed the `nullptr` check.

g-h-c commented on projects/hip/include/hip/hip_runtime_api.h:
```diff
@@ -6356,6 +6356,27 @@ hipError_t hipModuleGetFunction(hipFunction_t* function, hipModule_t module, con
+ * @param[in] dev Device to query attribute of
+ *
+ * @returns #hipSuccess, #hipErrorDeinitialized, #hipErrorNotInitialized, #hipErrorInvalidHandle,
+ #hipErrorInvalidValue, #hipErrorInvalidDevice
```
I think hipErrorMissingConfiguration is missing

sluzynsk-amd replied:
Now that I removed the nullptr check I think it should not return `hipErrorMissingConfiguration`. I copied the return values from cuda docs

g-h-c commented on projects/clr/hipamd/src/hip_library.cpp:
```diff
@@ -177,4 +178,62 @@ hipError_t hipLibraryGetKernel(hipKernel_t* kernel, hipLibrary_t library, const
+    case HIP_FUNC_ATTRIBUTE_CACHE_MODE_CA:
+      *pi = 0;
+      break;
+    case HIP_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES:
```
This is not taking into account wrkGrpInfo_->usedLDSSize_. CUDA does take into account how much shared memory the symbol uses when returning the value associated with CU_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES.

I reckon the calculation should be the same we are implementing in [hipOccupancyAvailableDynamicSMemPerBlock](https://github.com/ROCm/rocm-systems/pull/899) but for only one block.

sluzynsk-amd replied:
Changed the code to use `wrkGrpInfo.maxDynamicSharedSizeBytes_` and aligned it to 256

g-h-c replied:
It is not clear to me what happened with HIP_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES; if you plan to add it in a future PR, then please explain that in the description of this PR.

g-h-c commented on projects/hip-tests/catch/unit/library/hipKernelGetAttribute.cc:
```diff
@@ -0,0 +1,109 @@
+    HIP_CHECK_ERROR(
+        hipKernelGetAttribute(&pi, HIP_FUNC_ATTRIBUTE_BINARY_VERSION, nullptr, device_id),
+        hipErrorInvalidResourceHandle);
+  }
```
HIP_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES is not being tested

sluzynsk-amd replied:
Added a new test case

g-h-c commented on projects/clr/hipamd/src/hip_library.cpp:
```diff
@@ -30,6 +30,7 @@ THE SOFTWARE.
 #include "hip_platform.hpp"
 #include "utils/debug.hpp"
 
+
```
This extra blank line seems unnecessary

g-h-c commented on projects/clr/hipamd/src/hip_library.cpp:
```diff
@@ -177,4 +178,62 @@ hipError_t hipLibraryGetKernel(hipKernel_t* kernel, hipLibrary_t library, const
+      break;
+    case HIP_FUNC_ATTRIBUTE_MAX_DYNAMIC_SHARED_SIZE_BYTES: {
+      int maxDynamicSharedSizeBytes = static_cast<int>(wrkGrpInfo.maxDynamicSharedSizeBytes_);
+      constexpr size_t ALIGNMENT = 256;
```
I am not sure we should merge this. I think we need to wait for Dragoslav to merge the alignment column in the ISA table

sluzynsk-amd replied:
I removed it and will add it in later commit

g-h-c commented on projects/hip-tests/catch/unit/library/loadlib_co.cc:
```diff
@@ -56,7 +56,7 @@ TEST_CASE("Unit_hip_library_load_co") {
     unsigned int count = 0;
     HIP_CHECK(hipLibraryGetKernelCount(&count, library));
-    REQUIRE(count == 3);
+    REQUIRE(count == 4);
```
I think this check is best done outside the sections as its value does not depend on which section we are in

sluzynsk-amd replied:
This test wasn't created by me.  I just modified it to facilitate for my kernel with dynamic memory

g-h-c commented on projects/hip/include/hip/hip_runtime_api.h:
```diff
@@ -6361,6 +6361,27 @@ hipError_t hipModuleGetFunction(hipFunction_t* function, hipModule_t module, con
+ #hipErrorInvalidValue, #hipErrorInvalidDevice
+ *
+ * Returns in *pi the integer value of the attribute attrib for the kernel kernel for the requested
+ device dev. The supported attributes are:
```
The supported attributes are (...)?

sluzynsk-amd replied:
added :)

rakesroy commented:
Unit_hipKernelGetAttribute_Positive_Basic failed in stress testing of PSDB build Linux# 5574: 
[SWDEV-564518](https://ontrack-internal.amd.com/browse/SWDEV-564518)
[SWDEV-564512](https://ontrack-internal.amd.com/browse/SWDEV-564512)
[SWDEV-565311](https://ontrack-internal.amd.com/browse/SWDEV-565311)

Unit_hipKernelGetAttribute_Negative_Parameters failed in stress testing of PSDB build Linux# 5574: 
[SWDEV-564512](https://ontrack-internal.amd.com/browse/SWDEV-564512)

vstojilj committed:
Fix invalid use of hipEventElapsedTime (#2698)

litvaOo committed:
Revert "rocr: SVMPrefetch to a particular numa node (#1063)" (#2792)

This reverts commit 0ba5a01baa367c58361170521ed8f24dfc4798fb.

akolliasAMD committed:
added rocshmem-reviewers on CODEOWNERS file (#2809)

habajpai-amd committed:
[rocprof-sys] Fix RCCL comm_data counters in rocpd output (#2607)

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->
The validate-rccl-* tests were failing because "RCCL Comm" counters were not being written to perfetto traces when using the new cached-perfetto approach.

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->
Root Cause: The write_perfetto_counter_track() in rccl.cpp was only called when config::get_use_perfetto() returned true, which requires ROCPROFSYS_TRACE_LEGACY=ON. This meant RCCL counters weren't captured with the new trace cache approach.

Solution: Integrated RCCL with the trace cache system:

Changes to source/lib/rocprof-sys/library/rocprofiler-sdk/rccl.cpp:

- Added cache_rccl_comm_data_events<Track>() function to store RCCL comm data via pmc_event_with_sample with category::comm_data
- Modified tool_tracing_callback_rccl() to always cache events for new perfetto approach, while preserving legacy write_perfetto_counter_track() calls for backward compatibility

Changes to tests/rocprof-sys-testing.cmake:

- Added rccl_api to ROCPROFSYS_ROCM_DOMAINS to enable RCCL API callback tracing

Handler verification: The perfetto_processor_t already has a handler for ROCPROFSYS_CATEGORY_COMM_DATA in m_pmc_track_map that processes the cached events.

yugang-amd committed:
Fix link for global enum and defines (#2802)

nileshnegi committed:
Add rccl-reviewers to CODEOWNERS file (#2816)

* Add rccl-reviewers to CODEOWNERS file
* Re-order lines based on copilot suggestion

sputhala-amd committed:
Update CHANGELOG.md (#2813)

Modified the release version for the XGMI/PCIe feature.

habajpai-amd committed:
fix roctx range markers not paired correctly in rocpd output (#2793)

## Motivation

Fix roctx range markers (Push/Pop, Start/Stop) not being displayed correctly in rocpd output. The Visualizer was showing only Stop/Pop events as instant markers instead of proper duration ranges with labels, while Perfetto output displayed them correctly.

## Technical Details

In `tool_tracing_callback_stop()`, the rocpd/database output was using `user_data->value` (timestamp of the Pop/Stop event) instead of `begin_ts` (corrected timestamp from the corresponding Push/Start event) when calling `cache_region()`.

The Perfetto output already used `begin_ts` correctly (line 818). This change aligns the rocpd output with the Perfetto behavior by using `begin_ts` instead of `user_data->value` (line 887).

Updated rocpd validation rules

marantic-amd committed:
[rocprof-sys] Use fmt APIs to construct strings instead of JOIN (#2643)

## Motivation

With the introduction of the new logging system base on `spdlog` library, opportunity shows to replace `timemory` dependent JOIN implementation with `fmt` library `format` and `join` APIs, which are shipped as a part of `spdlog` lib

## Technical Details

Use `fmt` provided APIs to properly format and package strings.

marantic-amd committed:
[rocprof-sys] Align rocpd symbols of the same counter types (#2675)

## Motivation

In order for Optiq to be able to detect that counter tracks are of the same type, we aligned `info_pmc` symbol naming across the tracks of the same type. Being able to know this will be useful for grouping and categorizing similar types of counter tracks and for setting up a consistent y-axis scale when plotting the values on charts.

## Technical Details

Replace unique and/or ordered symbol names with counter-common symbol name which will be the same for the counters of the same type, with counter track name remaining the unique identifier for that counter track. For example, the "symbol" field was "JpegAct_0" but is now "JpegAct".

jamessiddeley-amd committed:
[rocprofiler-compute] Threshold Based Clamping in Analyze Stage (#2565)

* add threshold clamping function + parse in parser.py (with I/O)

* implemented hybrid threshold solution

* update changelog

* removed absolute threshold hybrid approach; restored relative threshold + warn

* edited warning msg, threshold -> 1%

* update changelog

* added 2 test cases

* ran master workflow yaml config files

* added to FAQ

* Revert "ran master workflow yaml config files"

This reverts commit 75a670e14d6f1619ebbda0ec218755ccbe0d22b1.

* update FAQ

* update config hashes

* Broke down long functions into Class with sub-functions

* ruff format

* addressed comments
